pkg/reconciler/kubernetes/tektoninstallerset/install.go

```diff
@@ -118,6 +132,123 @@ func (i *installer) EnsureDeploymentResources() error {
+	if err != nil {
+		return err
+	}
+
+	hash, err := computeDeploymentHash(*dep)
+	if err != nil {
+		return fmt.Errorf("failed to compute hash of deployment: %v", err)
+	}
+
+	if len(dep.Annotations) == 0 {
+		dep.Annotations = map[string]string{}
+	}
+	dep.Annotations[lastAppliedHashKey] = hash
+
+	unstrObj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(dep)
+	if err != nil {
+		return err
+	}
+	expected.SetUnstructuredContent(unstrObj)
+
+	return i.Manifest.Client.Create(expected)
+}
+
+func (i *installer) updateDeployment(existing *unstructured.Unstructured, existingDeployment, expectedDeployment *appsv1.Deployment) error {
+
+	// save on cluster replicas in a var and assign it back to deployment
+	onClusterReplicas := existingDeployment.Spec.Replicas
+
+	existingDeployment.Spec = expectedDeployment.Spec
+	existingDeployment.Spec.Replicas = onClusterReplicas
+
+	// compute new hash of spec and add as annotation
+	newHash, err := computeDeploymentHash(*existingDeployment)
+	if err != nil {
+		return fmt.Errorf("failed to compute new hash of existing deployment: %v", err)
+	}
+
+	if len(existingDeployment.Annotations) == 0 {
+		existingDeployment.Annotations = map[string]string{}
+	}
+
+	existingDeployment.Annotations[lastAppliedHashKey] = newHash
+
+	unstrObj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(existingDeployment)
+	if err != nil {
+		return err
+	}
+	existing.SetUnstructuredContent(unstrObj)
+
+	return i.Manifest.Client.Update(existing)
+}
+
+func (i *installer) ensureDeployment(expected *unstructured.Unstructured) error {
+
+	// check if deployment already exist
+	existing, err := i.Manifest.Client.Get(expected)
+	if err != nil {
+
+		// If deployment doesn't exist, then create new
+		if apierrs.IsNotFound(err) {
+			return i.createDeployment(expected)
+		}
+		return err
+	}
+
+	// if already exist then check if hash is changed
+	existingDeployment := &appsv1.Deployment{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(existing.Object, existingDeployment); err != nil {
+		return err
+	}
+
+	expectedDeployment := &appsv1.Deployment{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(expected.Object, expectedDeployment); err != nil {
+		return err
+	}
+
+	// compare spec hash with the one saved in annotation
+	// if annotation doesn't exist then update the deployment
+
+	existingDepSpecHash, err := computeDeploymentHash(*existingDeployment)
+	if err != nil {
+		return fmt.Errorf("failed to compute hash of existing deployment: %v", err)
+	}
+
+	hashFromAnnotation, hashExist := existingDeployment.Annotations[lastAppliedHashKey]
+
+	// if hash doesn't exist then update the deployment with hash
+	if !hashExist {
+		return i.updateDeployment(existing, existingDeployment, expectedDeployment)
+	}
+
+	// if both hash are same, nothing is changed and we can return
+	if existingDepSpecHash == hashFromAnnotation {
```

Comment:
From our discussion, i understand the operator logic behind this check, the comparisson of hash stored (create time) in the annotation of an existing deployment with its current hash. This will solve the specific problem of solving current `HPA` changing replicas situation. However, the more I think about it I am leaning towards an approach where we relax the `reset by operator` behaviour. 

I think in the long run we should let our end user (eg: platform engineer) or other on cluster robot agents the option to do change spec of an existing deployment. This means the operaotor will check the compute the hash of `expected hash` (to be applied) and compare it with the hash stored in the annotation of an existing deployment.

This means operator's objectives will become:
1. recreate the deployement if it does not exist
2. update (re-apply) new spec if there is a change which is coming from higher level reconciler (eg: tekton-pipeline, tekton-config...) or from an upgrade

and an operator `non-goal` will be:
1. reverting (resetting) a deployment back to the one the state operator created if an enduser (platform engineer) or an robot agent changes it.

Reply:
let us discuss this and try to reach a consensus 🧑‍💻 🧰

Reply:
cc @vdemeester @concaf @savitaashture @pradeepitm12